### ISSUE TITLE
[Bug]: `flatpak-spawn` uses different sandbox then the one it has been called from

### ISSUE THREAD

**Issue by sophie-h** [flatpak/flatpak#5973]
### Checklist

- [X] I agree to follow the [Code of Conduct](https://github.com/flatpak/flatpak/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.
- [X] I have searched the [issue tracker](https://www.github.com/flatpak/flatpak/issues) for a bug that matches the one I want to file, without success.
- [X] If this is an issue with a particular app, I have tried filing it in the appropriate issue tracker for the app (e.g. under https://github.com/flathub/) and determined that it is an issue with Flatpak itself.
- [X] This issue is not a report of a security vulnerability (see [here](https://github.com/flatpak/flatpak/blob/main/SECURITY.md) if you need to report a security issue).

### Flatpak version

1.15.10

### What Linux distribution are you using?

Debian

### Linux distribution version

testing

### What architecture are you using?

x86_64

### How to reproduce

1. Install two versions of an app, for example <id>//stable and <id>//test from Flathub
2. Use `flatpak-spawn` inside the sandbox
3. The command is potentially spawned inside a different sandbox/app version than is was started fromit

### Expected Behavior

The process is spawned in a sandbox that contains the same files and permissions as the one it was spawned from. Not a completely different one.

### Actual Behavior

It looks like the portal always takes the sandbox that `flatpak run <app>` would use, so the one that's set via `make-current`. But not 100% sure.

### Additional Information

Ideally, the Flatpak wouldn't even have to be installed for `flatpak-spawn` to work. Can't the portal remember what the command were with which the Flatpak was started originally and take these to spawn the new sandbox?

The current situation makes it very difficult to test apps that use sandboxing. Or is there a trick? For glycin, I'm disabling the sandbox in development environments which is not handy.

@mcatanzaro @GeorgesStavracas How does Epiphany handle this? How can the sandbox be spawned there in development Flatpaks?

**Comment by mcatanzaro** [comment]
> @mcatanzaro @GeorgesStavracas How does Epiphany handle this? How can the sandbox be spawned there in development Flatpaks?

I think it breaks. The WebKit version has to exactly match or all sorts of weird problems can occur.

**Comment by mcatanzaro** [comment]
I deleted my previous comment since I hadn't fully read yours yet.

WebKit doesn't do anything special here. It just assumes that flatpak-spawn uses the same sandbox that it's run from. [Here is the code.](https://github.com/WebKit/WebKit/blob/8d89e0e1050501362dbff68b0fb0ff61484e9eec/Source/WebKit/UIProcess/Launcher/glib/FlatpakLauncher.cpp#L50) If it uses the wrong runtime, then the odds of WebKit crashing would be pretty high.

I've heard rumors that it might break if the runtime is updated while Epiphany is running, but I don't think I've seen that happen myself.

**Comment by sophie-h** [comment]
But that means that Epiphany nightly has to be installed for being able to run it from something like Builder?

**Comment by GeorgesStavracas** [comment]
> But that means that Epiphany nightly has to be installed for being able to run it from something like Builder?

Yes. Ultimately the Flatpak portal runs `$ flatpak run  --command=<cmd> <appId>` with the arguments the caller passes to the portal. This presumes that the app is installed either in the user or the system installation.

**Comment by mcatanzaro** [comment]
Huh, I didn't know about this problem, but I guess so.

**Comment by GeorgesStavracas** [comment]
I don't have any ideas on how to deal with the Builder situation (how can the Flatpak portal deal with Flatpaks that, as far as Flatpak is aware, doesn't exist?). But for the specificity problem, I think the easy fix is make the Flatpak portal call `flatpak run [--system|--user] org.foo.bar/<arch>/<branch>` instead of just passing the appid and hoping for the best.

**Comment by sophie-h** [comment]
> I don't have any ideas on how to deal with the Builder situation

I was thinking of using
```
flatpak build [OPTION...] DIRECTORY [COMMAND [ARG...]]
```

which Builder uses to start the app. The portal uses `flatpak build DIRECTORY` instead of `flatpak run APP_ID`.

I think the `DIRECTORY` and the information to use `flatpak build` are already available *somewhere*. I can get both via

```
grep 'app-path=\|build=' /.flatpak-info
```

The app-path is the `DIRECTORY` and `build` is set to `true`.

> I think the easy fix is make the Flatpak portal call `flatpak run [--system|--user] org.foo.bar/<arch>/<branch>` instead of just passing the appid and hoping for the best.

Yeah, that's probably what it should have been to begin with xD

**Comment by GeorgesStavracas** [comment]
I think these two things should be handled separately. Improving the `flatpak run` command is immediately actionable, handling the Builder situation will require a lot more experimentation and thinking.

**Comment by sophie-h** [comment]
> I think these two things should be handled separately.

What do you mean by that? Do you want a second issue?

**Comment by GeorgesStavracas** [comment]
New information arrived: the Flatpak portal already specifies `--commit` and `--runtime-commit` when spawning. So it already does what I've mentioned.

This is how the portal spawns a new WebKit web process for GNOME Web:

```
/usr/bin/flatpak run --sandbox --socket=wayland --socket=fallback-x11 --socket=x11 --share=ipc --socket=pulseaudio --device=dri --session-bus --a11y-bus --env-fd=274 --parent-pid=4066650 --parent-expose-pids --unshare=network --filesystem=/home/feaneron/.var/app/org.gnome.Epiphany.Devel/data/epiphany:ro --filesystem=/home/feaneron/.var/app/org.gnome.Epiphany.Devel/cache/epiphany:ro --filesystem=/home/feaneron/.var/app/org.gnome.Epiphany.Devel/config/epiphany:ro --runtime=org.gnome.Platform --runtime-version=master --commit=a9f50796e60ce188bdefa06eb9072441c1e88a705e112deab189b12d8b03b463 --runtime-commit=c73b0235591b451cbf83275bd78cca152f6b33f50c142185fa0f126cb909b9c1 --cwd=/home/feaneron --command=/usr/libexec/webkitgtk-6.0/WebKitWebProcess org.gnome.Epiphany.Devel/x86_64/master 25 271 273
```

Now for the original question:

> flatpak-spawn uses different sandbox then the one it has been called from

Is Glycin using the `--sandbox` option? If so, then that's entirely expected as `--sandbox` is documented to drop all extra permissions and spawn the process in an airtight sandbox.

**Comment by sophie-h** [comment]
> New information arrived: the Flatpak portal already specifies `--commit`

So I guess it just omits that if it spawns from a build environment? Because with `build=true` there can't be any `app-commit`, right? Maybe I just noticed that Builder would select a kinda arbitrary version of the sandbox and I concluded that it's probably always the case.

> Is Glycin using the `--sandbox` option? If so, then that's entirely expected as `--sandbox` is documented to drop all extra permissions and spawn the process in an airtight sandbox.

Yes, I expect these permissions to be dropped. I meant different version with different sandbox.

**Comment by mcatanzaro** [comment]
Possibly related to: #5398 or #5570

**Comment by mcatanzaro** [comment]
@sophie-h do you still feel this is a flatpak bug? Does using `--commit` and `--runtime-commit` not work for you?
